src/controllers/request-breach-scan.js

```diff
@@ -0,0 +1,72 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { getMessage } from '../utils/fluent.js'
+import { getBreachLogo } from '../utils/breach-logo.js'
+import { UserInputError } from '../utils/error.js'
+import { getSha1 } from '../utils/fxa.js'
+import { getBreachesForEmail } from '../utils/hibp.js'
+
+/**
+ * @typedef {{ success: false }} RequestBreachScanErrorResponse
+ * @typedef {{ success: true, breaches: Array<unknown>, heading: string, logos: string[], dataClassStrings: string[][], total: number }} RequestBreachScanSuccessResponse
+ * @typedef {RequestBreachScanErrorResponse | RequestBreachScanSuccessResponse} RequestBreachScanResponse
+ */
+
+/** @type {import('express').RequestHandler<unknown, RequestBreachScanResponse>} */
+async function requestBreachScan (req, res, next) {
+  if (req.method !== 'POST' || typeof req.body?.email !== 'string') {
+    return next()
+  }
+
+  // TODO could share this validation logic with add-email API?
+  const email = req.body.email
+  // Use the same regex as HTML5 email input type
+  // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#basic_validation
+  const emailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/
+
+  if (!email || !emailRegex.test(email)) {
+    throw new UserInputError(getMessage('user-add-invalid-email'))
+  }
+
+  try {
+    const allBreaches = req.app.locals.breaches
+    const breaches = await getBreachesForEmail(getSha1(email), allBreaches, false)
+
+    /** @type {RequestBreachScanSuccessResponse} */
+    const successResponse = {
+      success: true,
+      breaches: breaches.slice(0, 6),
+      total: breaches.length,
+      heading:
+        // This is sent in the API response so we can replace the variables in
+        // the Fluent string (because Fluent might change the strings depending
+        // on the variables, specifically the count, and we don't run Fluent on
+        // the client side):
+        getMessage(
+          'exposure-landing-result-hero-heading',
+          {
+            email,
+            count: breaches.length
+          }
+        )
+          .replace('<email>', `<span class="breach-result-email" title="${email}">`)
```

Comment:
Per above, I'm not confident that this is sanitized that something couldn't sneak in here. Email addresses allow quite a range of characters, depending on sanitizing alone may not give us what we want.

One approach to handling potentially unsafe strings is to apply it using client-side DOM, there are a number of "safe sinks" https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#safe-sinks

If you want to stick to doing this on server-side, I'd at least run `encodeURIComponent(email)` but I'd suggest running it by security folks.

Reply:
Oh crap yes, nice catch. I didn't intend to inject it server-side, but then of couse I can into the issues with trying to inject the email address into a Fluent variable. !@#$%^&*(

I added bf99ddc08 to do it client-side and reuse the email that has been passed through the safe sink.